usage/src/org/apache/cloudstack/quota/QuotaAlertManagerImpl.java

```diff
@@ -0,0 +1,471 @@
+            return emailTemplateType;
+        }
+
+        public void sentSuccessfully() {
+            if (emailTemplateType == QuotaEmailTemplateTypes.QUOTA_STATEMENT) {
+                quotaAccount.setLastStatementDate(new Date());
+                _quotaAcc.update(quotaAccount.getAccountId(), quotaAccount);
+            }
+            else {
+                quotaAccount.setQuotaAlertDate(new Date());
+                quotaAccount.setQuotaAlertType(emailTemplateType.ordinal());
+                _quotaAcc.update(quotaAccount.getAccountId(), quotaAccount);
+            }
+        }
+    };
+
+    class EmailQuotaAlert {
+        private Session _smtpSession;
+        private final String _smtpHost;
+        private int _smtpPort = -1;
+        private boolean _smtpUseAuth = false;
+        private final String _smtpUsername;
+        private final String _smtpPassword;
+        private final String _emailSender;
+
+        public EmailQuotaAlert(String smtpHost, int smtpPort, boolean smtpUseAuth, final String smtpUsername, final String smtpPassword, String emailSender, boolean smtpDebug) {
+            _smtpHost = smtpHost;
+            _smtpPort = smtpPort;
+            _smtpUseAuth = smtpUseAuth;
+            _smtpUsername = smtpUsername;
+            _smtpPassword = smtpPassword;
+            _emailSender = emailSender;
+
+            if (_smtpHost != null) {
+                Properties smtpProps = new Properties();
+                smtpProps.put("mail.smtp.host", smtpHost);
+                smtpProps.put("mail.smtp.port", smtpPort);
+                smtpProps.put("mail.smtp.auth", "" + smtpUseAuth);
+                if (smtpUsername != null) {
+                    smtpProps.put("mail.smtp.user", smtpUsername);
+                }
+
+                smtpProps.put("mail.smtps.host", smtpHost);
+                smtpProps.put("mail.smtps.port", smtpPort);
+                smtpProps.put("mail.smtps.auth", "" + smtpUseAuth);
+                if (smtpUsername != null) {
+                    smtpProps.put("mail.smtps.user", smtpUsername);
+                }
+
+                if ((smtpUsername != null) && (smtpPassword != null)) {
+                    _smtpSession = Session.getInstance(smtpProps, new Authenticator() {
+                        @Override
+                        protected PasswordAuthentication getPasswordAuthentication() {
+                            return new PasswordAuthentication(smtpUsername, smtpPassword);
+                        }
+                    });
+                } else {
+                    _smtpSession = Session.getInstance(smtpProps);
+                }
+                _smtpSession.setDebug(smtpDebug);
+            } else {
+                _smtpSession = null;
+            }
+        }
+
+        public void sendQuotaAlert(List<String> emails, String subject, String body) throws MessagingException, UnsupportedEncodingException {
+            if (_smtpSession != null) {
```

Comment:
Consider simplifying this method by placing the following at the top of the method to simplify the flow and remove an unnecessary ``else`` block:
```
if (_smtpSession == null) {
    throw new RuntimeException("Unable to create smtp session");
}

// The rest of of the method minus the else block
```